File: lib/domain/cultivation_plan/dtos/field_cultivation_optimization_persist.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Dtos
      # FieldCultivation#optimization_result に保存するスナップショット（agrr allocation 由来）。
      class FieldCultivationOptimizationPersist
        attr_reader :allocation_id, :expected_revenue, :profit, :raw_allocation_document

        # @param allocation_id [Integer]
        # @param expected_revenue [Numeric]
        # @param profit [Numeric]
        # @param raw_allocation_document [Hash] agrr の allocation 行に相当する JSON 互換 Hash
        def initialize(allocation_id:, expected_revenue:, profit:, raw_allocation_document:)
          @allocation_id = allocation_id
          @expected_revenue = expected_revenue
          @profit = profit
          @raw_allocation_document = self.class.__send__(:deep_freeze_document, raw_allocation_document)
          freeze
        end

        # ActiveRecord serialize(JSON) 向けミュータブル Hash
        # @return [Hash]
        def to_storage_hash
          {
            allocation_id: allocation_id,
            expected_revenue: expected_revenue,
            profit: profit,
            raw: Marshal.load(Marshal.dump(raw_allocation_document))
          }
        end

        def self.deep_freeze_document(doc)
          case doc
          when nil then nil
          when Hash then doc.transform_values { |v| deep_freeze_document(v) }.freeze
          when Array then doc.map { |v| deep_freeze_document(v) }.freeze
          else
            doc
          end
        end
        private_class_method :deep_freeze_document
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/dtos/field_cultivation_optimization_persist.rs
//! Ruby: `Domain::CultivationPlan::Dtos::FieldCultivationOptimizationPersist`

use serde_json::{json, Map, Value};

use crate::weather_data::helpers::copy_and_deep_freeze;

/// Ruby: `Domain::CultivationPlan::Dtos::FieldCultivationOptimizationPersist`
#[derive(Debug, Clone)]
pub struct FieldCultivationOptimizationPersist {
    pub allocation_id: i64,
    pub expected_revenue: f64,
    pub profit: f64,
    raw_allocation_document: Option<Value>,
}

impl FieldCultivationOptimizationPersist {
    pub fn new(
        allocation_id: i64,
        expected_revenue: f64,
        profit: f64,
        raw_allocation_document: Value,
    ) -> Self {
        Self {
            allocation_id,
            expected_revenue,
            profit,
            raw_allocation_document: copy_and_deep_freeze(Some(raw_allocation_document)),
        }
    }

    pub fn raw_allocation_document(&self) -> Option<&Value> {
        self.raw_allocation_document.as_ref()
    }

    /// ActiveRecord serialize(JSON) 向けミュータブル Hash
    pub fn to_storage_hash(&self) -> Map<String, Value> {
        let mut map = Map::new();
        map.insert("allocation_id".into(), json!(self.allocation_id));
        map.insert("expected_revenue".into(), json!(self.expected_revenue));
        map.insert("profit".into(), json!(self.profit));
        map.insert(
            "raw".into(),
            copy_and_deep_freeze(self.raw_allocation_document.clone())
                .unwrap_or(Value::Null),
        );
        map
    }
}
